### DISCUSSION TITLE
[Feature]: Use branch name as subdomain

### DISCUSSION THREAD

**Discussion by martijnmichel** [coollabsio/coolify#3131] [Feature Requests]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

It would be great to be able to access the latest push to a branch through a named url on your coolify deploy. (just like vercel recently added)

Ie: having a branch name "coolify" would have the latest preview deploys of your repo at https://coolify.yourdomain.com or with generic name https://coolify-git.yourdomain.com https://coolify-anythingyouchoose.yourdomain.com

### Why should this be worked on?

It greatly improves team/customer work.

**Comment by unownone** [comment]
I think just branchname might get too vague as branchnames can easily conflict. `service 1` and `service 2` can always have the same branch `main` or `master`. Moreover, I think coolify deploys by branch name , so multi branch deploys may be something that need to be looked at differently right?

**Comment by martijnmichel** [comment]
I think i already saw a feature request for multiple domains on the same app/service, that would already help alot.

In our case the yourdomain.com part is already different for each app so there won't be any overlapping. But if not, it should be more generic with added appname part. ie: https://org-branch-repo.yourdomain.com next to org-commit-repo.yourdomain.com

**Comment by ChronSyn** [comment]
One approach I see elsewhere is to use the short commit id (e.g. 1a2b3c) followed by the branch name (but truncated to a maximum amount of characters) replacing any _ or space with - . e.g. `f234da-login-screen-refac`.

This would significantly reduce the chance of conflicts while also keeping preview domains relevant and easy to identify.

**Comment by Mobilpadde** [comment]
Is this about PRs or am I missing something? 

If so, I agree with @martijnmichel and the branch-name but along with the PR number, e.g. this PR #1095 could be called `v4-next-1` (`branch-PR#`) or whatever people finds useful.

Building on @ChronSyn's idea, a neat feature along with this would be the ability use variables:

* `branch`: The branch name
  * This could have a few varieties like: `branch{#}` whereas the pound-sign would be any number - Building on previous example: `branch{2}` = `v4` or `branch{6}` = `v4-nex`
* `pr`: The PR number
* `sha`: Defaulting to the complete SHA or like `branch{#}` with the cut. This would make the cut a generic feature, that'd be re-used muultiple times.

Taking the liberty to ping @andrasbacsai (GH needs a `@maintainers` ping :shrug:), as I think this'd be a great benefit for `v4`.

**Comment by martijnmichel** [comment]
I am actually talking about viewing commits on different urls. While actually this also includes a deploy of a PR (be it as latest commit into the branch that the PR is about).

For instance I recently made a small update on our starter template updating to next 13 in branch next13. It is avail on the following urls:
https://ecbase-react-starter-git-next13-ecbase.vercel.app/ ---using repo-branch-"git"-org
https://ecbase-react-starter-bo4gngfhv-ecbase.vercel.app/ ---using repo-commit-org

Would love to see this also in v4 as it makes finding/testing branches super easy accross teams.

**Comment by Mobilpadde** [comment]
> I am actually talking about viewing commits on different urls. While actually this also includes a deploy of a PR (be it as latest commit into the branch that the PR is about).
> 
> For instance I recently made a small update on our starter template updating to next 13 in branch next13. It is avail on the following urls:
> https://ecbase-react-starter-git-next13-ecbase.vercel.app/ ---using repo-branch-"git"-org
> https://ecbase-react-starter-bo4gngfhv-ecbase.vercel.app/ ---using repo-commit-org
> 
> Would love to see this also in v4 as it makes finding/testing branches super easy accross teams.

Ah, gotcha! 

But honestly I don't really see the use-case, now. I'm wondering, wouldn't branching out again solve that issue? This way it's also possible to make multiple changes that don't follow each other completely 🤔

**Comment by martijnmichel** [comment]
> > I am actually talking about viewing commits on different urls. While actually this also includes a deploy of a PR (be it as latest commit into the branch that the PR is about).
> > For instance I recently made a small update on our starter template updating to next 13 in branch next13. It is avail on the following urls:
> > https://ecbase-react-starter-git-next13-ecbase.vercel.app/ ---using repo-branch-"git"-org
> > https://ecbase-react-starter-bo4gngfhv-ecbase.vercel.app/ ---using repo-commit-org
> > Would love to see this also in v4 as it makes finding/testing branches super easy accross teams.
> 
> Ah, gotcha!
> 
> But honestly I don't really see the use-case, now. I'm wondering, wouldn't branching out again solve that issue? This way it's also possible to make multiple changes that don't follow each other completely 🤔

Sorry, I've explained wrongly, it's about branches. In those url's the first one holds all the latest **commit**s to that branch. That way, while developing on a branch, you can always go to that same url (or give it to your clients) and it will be your latest dev code in that branch.

**Comment by Mobilpadde** [comment]
> > > I am actually talking about viewing commits on different urls. While actually this also includes a deploy of a PR (be it as latest commit into the branch that the PR is about).
> > > For instance I recently made a small update on our starter template updating to next 13 in branch next13. It is avail on the following urls:
> > > https://ecbase-react-starter-git-next13-ecbase.vercel.app/ ---using repo-branch-"git"-org
> > > https://ecbase-react-starter-bo4gngfhv-ecbase.vercel.app/ ---using repo-commit-org
> > > Would love to see this also in v4 as it makes finding/testing branches super easy accross teams.
> > 
> > 
> > Ah, gotcha!
> > But honestly I don't really see the use-case, now. I'm wondering, wouldn't branching out again solve that issue? This way it's also possible to make multiple changes that don't follow each other completely thinking
> 
> Sorry, I've explained wrongly, it's about branches. In those url's the first one holds all the latest **commit**s to that branch. That way, while developing on a branch, you can always go to that same url (or give it to your clients) and it will be your latest dev code in that branch.

Is that to "hide" the new changes from a client? Otherwise I don't see a need to change the URL every commit :thinking: I mean, the URL will always change on each commit :shrug:

Should old URLs be saved for history purposes, or always generate a new?

**Comment by luckydonald** [comment]
To chime in on that, I'm looking to use a branch without the need to create a PR.

For Vercel it's basically `${project_name}-${branch_name}-${username}`. This one is the same for all commits of that branch.
Additionally to single commit subdomain `${project_name}-${commit}-${username}`. This one will always change with the commits.

**Comment by Mobilpadde** [comment]
> To chime in on that, I'm looking to use a branch without the need to create a PR.
> 
> For Vercel it's basically `${project_name}-${branch_name}-${username}`. This one is the same for all commits of that branch.
> Additionally to single commit subdomain `${project_name}-${commit}-${username}`. This one will always change with the commits.
> 

Ah, yeah. I agree, my old comments on PRs should be read as branches. I'm just horrible at explaining ~~sometimes~~ mostly 🙈

**Comment by luckydonald** [comment]
I feel like it could be settings like this, allowing you to set the subdomains dynamically:

---

**Deployment**
_You can use the variables `${branch}`, `${commit}`, `${project}`, `${pr}`, `${username}`._
- [x] Deploy main branch: <kbd> cool-custom-domain.tld </kbd>
- [x] Deploy pull-requests branch: <kbd> ${pr}-${project}.example.com </kbd>
- [x] Deploy branches: <kbd> ${branch}-${project}.example.com </kbd>

**Comment by luckydonald** [comment]
@peaklabs-dev why was this downgraded from an issue (feature request) to a discussion?

**Reply by peaklabs-dev** [reply]
Because Issues is now only for bugs. Improvements and feature requests are now manged in discussions.

**Reply by luckydonald** [reply]
@peaklabs-dev does that mean this is no longer planned?

I've seen [duplicate requests](https://github.com/coollabsio/coolify/issues/4680) for this pop up because the discussions are not really visible...

**Reply by hacknug** [reply]
It just means issues are used for bugs on the current codebase and discussions for ideas and feature requests such as this one. I'm not sure this was ever planned but it was at least suggested.
